languages/python/atom/element.py

```
@@ -3530,24 +3527,39 @@ def _reference_get_shared_mem(
-                        # system
-                        shape_tuple = ()
-                        for dim in (width, height, depth):
-                            if dim != 0:
-                                shape_tuple = shape_tuple + (dim,)
-                            else:
-                                break
-
-                        data = data.reshape(shape_tuple)
-                        data.dtype = np.uint8
-                    except Exception as e:
+                    except atomshmem.AtomShMemException as e:
                         self.logger.warning(
                             f"An exception occurred when getting data for handle {handle}: {str(e)}"
                         )
+                        continue
+
+                    # need to check the shape of the return data because
+                    # if it is < 3D then the reshape command will throw
+                    # a warning for the dimensions that are "0" in the
+                    # system
+                    shape_tuple = ()
+                    for dim in (width, height, depth):
+                        if dim != 0:
+                            shape_tuple = shape_tuple + (dim,)
+                        else:
+                            break
+
+                    # reshape the data, but can't use "reshape" function
+                    # because that copies the data
+                    try:
+                        data.shape = shape_tuple
+                    except ValueError as e:
+                        self.logger.warning(
+                            f"Cannot reshape data as the total size of the \
+                                array must be unchanged: {str(e)}"
+                        )
+                    except AttributeError as e:
+                        self.logger.warning(
+                            f"Cannot perform inplace modification of data \
+                                because of incompatible shape: {str(e)}"
+                        )
```

So the behavior here in case the reshape fails is to return the data in its original shape, is that right? Could there be any issues then when we try to use it, or do we think generally it will be ok? Mainly wondering if its better to return None when we can't do the reshape, I'm not sure

---

If we can't reshape I think something went wrong in our internal implementation of shared memory -- it shouldn't happen and should likely raise?

---

When a GET fails for some reason we return None for that key instead of raising since we may be getting multiple references per call. But yeah I'm also thinking we should probably indicate it failed instead of just returning the raw data.

---

That's slightly different as there's an external dependency on Redis which can fail. We have an internal dependency on our own code here which should be rock solid, so it makes more sense to raise/yell loudly if this happens since it really shouldn't by design.

---

So if a user tries to get multiple references in a single call and one datum fails to reshape they should not receive the data for the other keys? I think its preferable that they still get the good data to continue working with, since there's no reason to fail on all keys if just one is causing an issue.

---

I think we're maybe not on the same page about the root cause of the reshape failing -- to me this can't happen unless something is very wrong in our implementation. Once we've written/tested the code there's really no reason this error should happen, so my preference is to have this sort of thing yell loudly and fail since something is very wrong.

---

If it's an error that can happen in production due to things out of our control (redis being down, etc.) I would definitely agree with your strategy, but this seems like a core component of Atom that just shouldn't break.

---

I don't know if its possible in production but think its better to plan for the worst in case it is. Our tests would catch it failing regardless of how we let it fail.

---

I think we may just disagree here which is fine; I'm looking for the implementation to guarantee this can't happen which should be possible. I don't really agree that we should be error-handling around things which would be bugs in our own code. 

We can take it however you'd like and it's not a huge deal, just wanted to explain the philosophy.

---

That makes sense, I'm also fine with either way. I guess I don't feel familiar enough with shared memory or Iouri's library to be confident in what fail cases are or are not possible to unexpectedly hit in the field but if we're very certain this one is impossible then I feel ok with letting it raise.

---

I guess I see it as:
- We have a numpy array of a certain shape on the way in
- We store the shape and the array in shared mem
- We get the array back from shared mem and apply the stored shape

Not sure there are a lot of paths where it could/should fail

---

Yeah, you're right it should always work since even if the shape in the header was corrupted or something, its used to get the data so should always match the shape that's returned.

---

I'm not sure I follow the last part, but also I think it's preferable where, in the event the header is corrupted, this code raises rather than prints a warning. To me, there's something very wrong and our whole app is in a lot of trouble if shared memory is getting corrupted.